### PR TITLE
Improve checking in coords constructor

### PR DESCRIPTION
**Pull request by dschwoerer**

### CODE REVIEW

File: src/mesh/coordinates.cxx [hunk 1]
```diff
@@ -1964,3 +1878,107 @@ Field2D Coordinates::Laplace_perpXY(MAYBE_UNUSED(const Field2D& A),
...
+
+void Coordinates::checkCovariant() {
+  // Diagonal metric components should be finite
+  bout::checkFinite(g_11, "g_11", "RGN_NOCORNERS");
+  bout::checkFinite(g_22, "g_22", "RGN_NOCORNERS");
+  bout::checkFinite(g_33, "g_33", "RGN_NOCORNERS");
+  if (g_11.hasParallelSlices() && &g_11.ynext(1) != &g_11) {
+    for (int dy = 1; dy <= localmesh->ystart; ++dy) {
+      for (auto& sign : {1, -1}) {
+        bout::checkFinite(g_11.ynext(sign * dy), "g_11.ynext",
```
[the remark this answers is not in the document]
Comment (ZedThree) [reply src/mesh/coordinates.cxx:1890]: Aside from the suggestion being on the wrong line, this should presumably be `const auto sign` anyway

File: src/mesh/coordinates.cxx [hunk 2]
```diff
@@ -1964,3 +1878,107 @@ Field2D Coordinates::Laplace_perpXY(MAYBE_UNUSED(const Field2D& A),
...
 }
+
+void Coordinates::checkCovariant() {
+  // Diagonal metric components should be finite
+  bout::checkFinite(g_11, "g_11", "RGN_NOCORNERS");
+  bout::checkFinite(g_22, "g_22", "RGN_NOCORNERS");
+  bout::checkFinite(g_33, "g_33", "RGN_NOCORNERS");
+  if (g_11.hasParallelSlices() && &g_11.ynext(1) != &g_11) {
+    for (int dy = 1; dy <= localmesh->ystart; ++dy) {
+      for (auto& sign : {1, -1}) {
```
Comment (ZedThree) [inline src/mesh/coordinates.cxx:1889]:
```suggestion
      for (const auto sign : {1, -1}) {
```

File: src/mesh/coordinates.cxx [hunk 3]
```diff
@@ -1964,3 +1878,107 @@ Field2D Coordinates::Laplace_perpXY(MAYBE_UNUSED(const Field2D& A),
...
+      }
+    }
+  }
+  // Diagonal metric components should be positive
+  bout::checkPositive(g_11, "g_11", "RGN_NOCORNERS");
+  bout::checkPositive(g_22, "g_22", "RGN_NOCORNERS");
+  bout::checkPositive(g_33, "g_33", "RGN_NOCORNERS");
+  if (g_11.hasParallelSlices() && &g_11.ynext(1) != &g_11) {
+    for (int dy = 1; dy <= localmesh->ystart; ++dy) {
+      for (auto& sign : {1, -1}) {
```
Comment (ZedThree) [inline src/mesh/coordinates.cxx:1905]:
```suggestion
      for (const auto sign : {1, -1}) {
```

File: src/mesh/coordinates.cxx [hunk 4]
```diff
@@ -1964,3 +1878,107 @@ Field2D Coordinates::Laplace_perpXY(MAYBE_UNUSED(const Field2D& A),
...
+    }
+  }
+
+  // Off-diagonal metric components should be finite
+  bout::checkFinite(g_12, "g_12", "RGN_NOCORNERS");
+  bout::checkFinite(g_13, "g_13", "RGN_NOCORNERS");
+  bout::checkFinite(g_23, "g_23", "RGN_NOCORNERS");
+  if (g_23.hasParallelSlices() && &g_23.ynext(1) != &g_23) {
+    for (int dy = 1; dy <= localmesh->ystart; ++dy) {
+      for (auto& sign : {1, -1}) {
```
Comment (ZedThree) [inline src/mesh/coordinates.cxx:1922]:
```suggestion
      for (const auto sign : {1, -1}) {
```

**Review by ZedThree — CHANGES_REQUESTED** [review]
Thanks @dschwoerer, I'm always in favour of obvious helper functions like this! Just some small changes to make clang-tidy happy

**Comment by dschwoerer** [comment]
Thanks! I was wondering whether clang-tidy would post a corrected version of the suggestion (and then forgot) :+1:

**Comment by dschwoerer** [comment]
734 conflicts, over 10k changed lines, over 200 files changed - that used to be a small change. Thanks clang-tidy!

I have rebased (removing clang-tidy commits) and force pushed - maybe it works now ...

File: include/bout/coordinates.hxx [hunk 5]
```diff
@@ -235,6 +235,9 @@ private:
   /// Set the parallel (y) transform from the options file.
   /// Used in the constructor to create the transform object.
   void setParallelTransform(Options* options);
+
+  void checkCovariant();
+  void checkContravariant();
```
Comment (ZedThree) [inline include/bout/coordinates.hxx:240]: Please can you just put a docstring on these? I know it's sort of obvious from the name, but checking _what_ would be good to know

**Review by ZedThree — CHANGES_REQUESTED** [review]

File: include/bout/coordinates.hxx (hunk 5 above)
Comment (dschwoerer) [reply include/bout/coordinates.hxx:240]:
```suggestion
  // check that covariant tensors are positive (if expected) and finite (always)
  void checkCovariant();
  // check that contravariant tensors are positive (if expected) and finite (always)
  void checkContravariant();
```

**Review by ZedThree — APPROVED** [review]

**Pull request merged**